src/replication.c

```diff
@@ -1630,38 +1634,93 @@ int replicationApply(struct raft *r)
     return rv;
 }
 
-void replicationQuorum(struct raft *r, const raft_index index)
+raft_index startIndexToCommit(const struct raft *r)
 {
-    size_t votes = 0;
-    size_t i;
-
-    assert(r->state == RAFT_LEADER);
+    unsigned i;
+    unsigned server_index;
+    raft_index idx;
+    raft_index max;
 
-    if (index <= r->commit_index) {
-        return;
+    /* In a single voter cluster where this node is part of the configuration,
+     * is a voter all stored entries can be committed. */
```

Comment:
Thanks, will investigate if it can be dropped.

Reply:
> Thanks, will investigate if it can be dropped.

Not sure it's clear, I mean that the ***words*** `is a voter` can be dropped from the comment (probably they are a leftover? the full comment does quite sound grammatically correct to me).

Reply:
Ah okay, yes the sentence indeed looks a bit strange.

Reply:
tests also still pass when I remove `r->configuration.servers[server_index].role == RAFT_VOTER` from the check, I will investigate anyway to make sure my understanding is correct.

Reply:
> tests also still pass when I remove `r->configuration.servers[server_index].role == RAFT_VOTER` from the check, I will investigate anyway to make sure my understanding is correct.

I was a bit suspicious about that condition too, but didn't mention it actually. I would not expect that cases where there is a single server and that server is not a voter to be possible, because otherwise there would be no way to make progress and commit further entries. So that condition could be turned into an assertion. However, I'm just hand-waving, not sure.

Reply:
The check actually says that there is 1 voter in the cluster, not that there is a single server. Checking for `r->configuration.servers[server_index].role == RAFT_VOTER` then means that this server was not demoted to standby or spare in the latest configuration.

Reply:
Ah yeah sorry, I misread the `configurationVoterCount` call, that's the number of voters indeed, not servers. Makes sense then.

Reply:
Ok, I double-checked and I think the condition is fine. The test still succeed after removing the `RAFT_VOTER` condition because the test is less strict then and the `start_index` is chosen to be `last_stored` on the current leader, which will be the highest possible index that can be committed anyway.